20 test cases pass, 16 test cases failed
----------------------------------------------------------------------
Ran 36 tests in 3.482s
FAILED (errors=16)

1） OSError: Unsupported cast dtype %dfloat64
 
2） OSError: synGraphCompile() logical failed = 26 eager = 1
ERROR: test_check_output_10 (__main__.TestLogicalOpComputation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/ddt.py", line 221, in wrapper
    return func(self, *args, **kwargs)
  File "/workspace_pdpd/repo/PaddleCustomDevice/backends/intel_hpu/tests/unittests/test_logical_op.py", line 124, in test_check_output
    self.check_output_hpu_with_cpu(self.logical_forward)
  File "/workspace_pdpd/repo/PaddleCustomDevice/backends/intel_hpu/tests/unittests/api_base.py", line 85, in check_output_hpu_with_cpu
    hpu_out = self.calc_result(forward, "intel_hpu")
  File "/workspace_pdpd/repo/PaddleCustomDevice/backends/intel_hpu/tests/unittests/api_base.py", line 63, in calc_result
    return calculator()
  File "/workspace_pdpd/repo/PaddleCustomDevice/backends/intel_hpu/tests/unittests/test_logical_op.py", line 110, in logical_forward
    return self.logical_api(x)
  File "/usr/local/lib/python3.10/dist-packages/paddle/tensor/logic.py", line 337, in logical_not
    return _C_ops.logical_not(x)
OSError: synGraphCompile() logical failed = 26 eager = 1
  [/workspace_pdpd/repo/PaddleCustomDevice/backends/intel_hpu/kernels/hpu_operator.cc:46]

3）`logical_xor` is not registered 
RuntimeError: (NotFound) The kernel with key (intel_hpu, Undefined(AnyLayout), float16) of kernel `logical_xor` is not registered and fail to fallback to CPU one. Selected wrong Backend `intel_hpu`. Paddle support following Backends: CPU.
  [Hint: Expected kernel_iter != iter->second.end(), but received kernel_iter == iter->second.end().] (at /paddle/paddle/phi/core/kernel_factory.cc:387)

--- a/backends/intel_hpu/tests/unittests/api_base.py
+++ b/backends/intel_hpu/tests/unittests/api_base.py
@@ -1,0 +1,129 @@
+# Copyright (c) 2025 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import paddle
+import numpy as np
+import unittest
+
+
+class TestAPIBase(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        paddle.seed(2066)
+        np.random.seed(2099)
+        paddle.disable_static()
+
+    def generate_bool_data(self, shape):
+        return np.random.randint(2, size=shape) == 1
+
+    def generate_float_data(self, shape, dtype):
+        return np.random.random(shape).astype(dtype)
+
+    def generate_integer_data(self, shape, dtype, lownum=0, highnum=256):
+        return np.random.randint(lownum, high=highnum, size=shape, dtype=dtype)
+
+    def generate_data(self, shape, dtype):
+        if dtype == np.float32 or dtype == np.float16 or dtype == np.float64:
+            data = self.generate_float_data(shape, dtype)
+        elif dtype == bool:
+            data = self.generate_bool_data(shape)
+        elif dtype == np.int8:
+            data = self.generate_integer_data(shape, dtype, -128, 127)
+        elif dtype == np.int16:
+            data = self.generate_integer_data(shape, dtype, -32768, 32767)
+        elif dtype == np.int32:
+            data = self.generate_integer_data(shape, dtype, -2147483648, 2147483647)
+        elif dtype == np.int64:
+            data = self.generate_integer_data(shape, dtype, -2147483648, 2147483647)
+        elif dtype == np.uint8:
+            data = self.generate_integer_data(shape, dtype, 0, 256)
+        elif dtype == np.uint16:
+            data = self.generate_integer_data(shape, dtype, 0, 65535)
+        elif dtype == np.uint32:
+            data = self.generate_integer_data(shape, dtype, 0, 4294967295)
+        elif dtype == np.uint64:
+            data = self.generate_integer_data(shape, dtype, 0, 4294967295)
+        else:
+            raise Exception("Unsupportted datatype!")
+        return data
+
+    def calc_result(self, calculator, device):
+        paddle.set_device(device)
+        return calculator()
+
+    def check_value(self, hpu_val, cpu_val, rtol, atol, label="output"):
+        self.assertTrue(
+            np.allclose(hpu_val, cpu_val, rtol, atol),
+            f"Compare hpu {label} with cpu failed.\n  hpu_{label}:{hpu_val}\n  cpu_{label}:{cpu_val}",
+        )
+
+    def check_multi_values(self, hpu_vals, cpu_vals, rtol, atol, label="output"):
+        self.assertTrue(
+            len(hpu_vals) == len(cpu_vals),
+            f"Compare hpu {label} with hpu failed.\n  hpu_{label}s:{hpu_vals}\n  cpu_{label}s:{cpu_vals}",
+        )
+
+        for hpu_val, cpu_val in zip(hpu_vals, cpu_vals):
+            self.assertTrue(
+                np.allclose(hpu_val, cpu_val, rtol, atol),
+                f"Compare hpu {label} with cpu failed.\n  hpu_{label}s:{hpu_vals}\n  cpu_{label}s:{cpu_vals}",
+            )
+
+    def check_output_hpu_with_cpu(self, forward, rtol=1e-5, atol=1e-5):
+        self.prepare_data()
+        hpu_out = self.calc_result(forward, "intel_hpu")
+        cpu_out = self.calc_result(forward, "cpu")
+        # print(f"hpu_out:{hpu_out}")
+        # print(f"cpu_out:{cpu_out}")
+        if isinstance(hpu_out, (list, tuple)):
+            self.check_multi_values(hpu_out, cpu_out, rtol, atol)
+        else:
+            self.check_value(hpu_out, cpu_out, rtol, atol)
+
+    def check_output_hpu_with_customized(
+        self, forward, customized, rtol=1e-5, atol=1e-5
+    ):
+        self.prepare_data()
+        hpu_out = self.calc_result(forward, "intel_hpu")
+        cpu_out = self.calc_result(customized, "cpu")
+        # print(f"hpu_out:{hpu_out}")
+        # print(f"cpu_out:{cpu_out}")
+        if isinstance(hpu_out, (list, tuple)):
+            self.check_multi_values(hpu_out, cpu_out, rtol, atol)
+        else:
+            self.check_value(hpu_out, cpu_out, rtol, atol)
+
+    def check_grad_hpu_with_cpu(self, backward, rtol=1e-5, atol=1e-5):
+        self.prepare_data()
+        hpu_grad = self.calc_result(backward, "intel_hpu")
+        cpu_grad = self.calc_result(backward, "cpu")
+        # print(f"hpu_grad:{hpu_grad}")
+        # print(f"cpu_grad:{cpu_grad}")
+        if isinstance(hpu_grad, (list, tuple)):
+            self.check_multi_values(hpu_grad, cpu_grad, rtol, atol, "grad")
+        else:
+            self.check_value(hpu_grad, cpu_grad, rtol, atol, "grad")
+
+    def check_grad_hpu_with_customized(
+        self, backward, customized, rtol=1e-5, atol=1e-5
+    ):
+        self.prepare_data()
+        hpu_grad = self.calc_result(backward, "intel_hpu")
+        cpu_grad = self.calc_result(customized, "cpu")
+        # print(f"hpu_grad:{hpu_grad}")
+        # print(f"cpu_grad:{cpu_grad}")
+        if isinstance(hpu_grad, (list, tuple)):
+            self.check_multi_values(hpu_grad, cpu_grad, rtol, atol, "grad")
+        else:
+            self.check_value(hpu_grad, cpu_grad, rtol, atol, "grad")

--- a/backends/intel_hpu/tests/unittests/test_logical_op.py
+++ b/backends/intel_hpu/tests/unittests/test_logical_op.py
@@ -1,0 +1,149 @@
+# Copyright (c) 2025 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function, division
+
+import unittest
+
+import numpy as np
+import paddle
+
+from ddt import ddt, data, unpack
+from api_base import TestAPIBase
+
+paddle.disable_static()
+
+import os
+from util import enable_paddle_static_mode
+
+intel_hpus_static_mode = os.environ.get(
+    "FLAGS_static_mode_intel_hpus", 0
+)  # default is dynamic mode test FLAGS_static_mode_intel_hpus=0
+
+LOGICAL_CASE = [
+    {"logical_api": paddle.logical_and, "shape": [2, 3, 32, 32], "dtype": np.float32},
+    {"logical_api": paddle.logical_and, "shape": [2, 3, 32, 32], "dtype": np.int32},
+    {"logical_api": paddle.logical_and, "shape": [2, 3, 32, 32], "dtype": np.int64},
+    {"logical_api": paddle.logical_and, "shape": [2, 4, 4], "dtype": np.float32},
+    {"logical_api": paddle.logical_and, "shape": [2, 3, 32, 32], "dtype": bool},
+    {"logical_api": paddle.logical_and, "shape": [2, 4, 4], "dtype": bool},
+    {"logical_api": paddle.logical_not, "shape": [2, 3, 32, 32], "dtype": np.float32},
+    {"logical_api": paddle.logical_not, "shape": [2, 3, 32, 32], "dtype": np.int32},
+    {"logical_api": paddle.logical_not, "shape": [2, 3, 32, 32], "dtype": np.int64},
+    {"logical_api": paddle.logical_not, "shape": [2, 4, 4], "dtype": np.float32},
+    {"logical_api": paddle.logical_not, "shape": [2, 3, 32, 32], "dtype": bool},
+    {"logical_api": paddle.logical_not, "shape": [2, 4, 4], "dtype": bool},
+    {"logical_api": paddle.logical_or, "shape": [2, 3, 32, 32], "dtype": np.float32},
+    {"logical_api": paddle.logical_or, "shape": [2, 3, 32, 32], "dtype": np.int32},
+    {"logical_api": paddle.logical_or, "shape": [2, 3, 32, 32], "dtype": np.int64},
+    {"logical_api": paddle.logical_or, "shape": [2, 4, 4], "dtype": np.float32},
+    {"logical_api": paddle.logical_or, "shape": [2, 3, 32, 32], "dtype": bool},
+    {"logical_api": paddle.logical_or, "shape": [2, 4, 4], "dtype": bool},
+    {"logical_api": paddle.logical_xor, "shape": [2, 3, 32, 32], "dtype": np.float32},
+    {"logical_api": paddle.logical_xor, "shape": [2, 3, 32, 32], "dtype": np.int32},
+    {"logical_api": paddle.logical_xor, "shape": [2, 3, 32, 32], "dtype": np.int64},
+    {"logical_api": paddle.logical_xor, "shape": [2, 4, 4], "dtype": np.float32},
+    {"logical_api": paddle.logical_xor, "shape": [2, 3, 32, 32], "dtype": bool},
+    {"logical_api": paddle.logical_xor, "shape": [2, 4, 4], "dtype": bool},
+]
+
+LOGICAL_F16_CASE = [
+    {
+        "logical_api": paddle.logical_and,
+        "numpy_api": np.logical_and,
+        "shape": [2, 3, 32, 32],
+        "dtype": np.float16,
+    },
+    {
+        "logical_api": paddle.logical_not,
+        "numpy_api": np.logical_not,
+        "shape": [2, 3, 32, 32],
+        "dtype": np.float16,
+    },
+    {
+        "logical_api": paddle.logical_or,
+        "numpy_api": np.logical_or,
+        "shape": [2, 3, 32, 32],
+        "dtype": np.float16,
+    },
+    {
+        "logical_api": paddle.logical_xor,
+        "numpy_api": np.logical_xor,
+        "shape": [2, 3, 32, 32],
+        "dtype": np.float16,
+    },
+]
+
+
+class TestLogicalOp(TestAPIBase):
+    def setUp(self):
+        self.init_attrs()
+        self.set_hpu()
+
+    def set_hpu(self):
+        enable_paddle_static_mode(int(intel_hpus_static_mode))
+
+    def init_attrs(self):
+        self.shape = [4]
+        self.dtype = np.float32
+
+    def prepare_data(self):
+        self.data_x = self.generate_data(self.shape, self.dtype)
+        self.data_y = self.generate_data(self.shape, self.dtype)
+
+        if self.dtype == np.int64 or self.dtype == np.int32:
+            self.data_x = self.generate_integer_data(self.shape, self.dtype, -200, 200)
+            self.data_y = self.generate_integer_data(self.shape, self.dtype, -200, 200)
+
+    def logical_forward(self):
+        x = paddle.to_tensor(self.data_x, dtype=self.dtype)
+        y = paddle.to_tensor(self.data_y, dtype=self.dtype)
+        if self.logical_api in [paddle.logical_not]:
+            return self.logical_api(x)
+        return self.logical_api(x, y)
+
+    def init_dtype(self):
+        self.dtype = np.float32
+
+
+@ddt
+class TestLogicalOpComputation(TestLogicalOp):
+    @data(*LOGICAL_CASE)
+    @unpack
+    def test_check_output(self, logical_api, shape, dtype):
+        self.logical_api = logical_api
+        self.shape = shape
+        self.dtype = dtype
+        self.check_output_hpu_with_cpu(self.logical_forward)
+
+
+@ddt
+class TestLogicalOpComputationF16(TestLogicalOp):
+    def numpy_binary(self):
+        if self.numpy_api in [np.logical_not]:
+            return self.numpy_api(self.data_x)
+        return self.numpy_api(self.data_x, self.data_y)
+
+    @data(*LOGICAL_F16_CASE)
+    @unpack
+    def test_check_output(self, logical_api, numpy_api, shape, dtype):
+        self.logical_api = logical_api
+        self.numpy_api = numpy_api
+        self.shape = shape
+        self.dtype = dtype
+        self.check_output_hpu_with_customized(self.logical_forward, self.numpy_binary)
+
+
+if __name__ == "__main__":
+    unittest.main()

